doc/auth_design.md

```diff
@@ -0,0 +1,104 @@
+cluster. With authentication and authorization, we will be able to implement *sessions*, which means that each SQL statement in a SQL program might be handled by different SQLFlow server instances in the Kubernetes service; however, the user wouldn't notice that.
+
+Authorization is not a too much a challenge because we can rely on
+SQL engines and training clusters, which denies requests if the user
+have no access.  In this document, we focus on authentication of SQLFlow users.
+
+## Design
+
+To make it modulized and extensible, we prefer to introduce an authentication server, a.k.a., auth server. We use a
+[Django](https://www.djangoproject.com/) Web server so that the authentication methods
+can extend to:
+
+- Database authentication
+- LDAP
+- User-defined authentication methods
+
+### Session
+
+A server-side "session" is needed to store credentials for each client to access
+the database and submitting jobs. The session can be defined as:
+
+```go
+type Session struct {
+    Token          int64  // useful only in "side-car" design
+    ClientEndpoint string // ip:port from the client
+    DBConnStr      string // mysql://127.0.0.1:3306
+    AK             string // access key
+    SK             string // secret key
+}
+```
+
+The token will act as the unique id of the session. The session object
+should be expired within some time and deleted on the server memory.
+
+We want to make sure that SQLFlow servers are stateless so that we can
+deploy it on any cluster that does auto fail-over and auto-scaling. In
+that case, we store session data into a reliable storage service like
+[etcd](https://github.com/etcd-io/etcd). 
+
+Possible two implementations listed below can satisfy what SQLFlow needs:
+
+### Authentication of SQLFlow Server
+
+**Note:** that SQLFlow should be dealing with three kinds of services:
+
+- SQLFlow service itself
+- Database service that stores the training data
+- A training cluster that runs the SQLFlow training job, e.g. Kubernetes
+
+SQLFlow should depend on the [SSO](https://en.wikipedia.org/wiki/Single_sign-on)
+service. Databases and training clusters also need to check
+if the user is valid and check if the user has granted proper permissions,
+but these services may have different credentials other than the SSO service.
+So there **must** be an "Auth Server" to fetch/create the user's AK/SK (access key/secret key)
+which will be used by databases or Kubernetes.
+
+For one case that we use MySQL as the database engine, the fetched AK/SK should
+be the MySQL's user and password. When running on the cloud environment, AK/SK
+should be the real user's keys.
+
+<img src="figures/sqlflow_auth.png">
+
+Users can use SQLFlow server with a simple jupyter notebook for simple deployment,
+for production deployments, users can take advantage of the cloud web IDE. The web
+IDE will redirect a user to the SSO service if the user is not logged in.
```

Comment:
Why would "the web IDE" redirect a user to the SSO service"? Is it configured to do so? Could users use Jupyter Notebook as their "web IDE"? If so, how should they configure it to work with SSO? And, how comes the SSO service? Who is supposed to build it?

Reply:
Removed all "web IDE" stuff and move to "JupyterHub"